src/basic/in-addr-util.c

```diff
@@ -179,44 +179,89 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
         /* Increases the network part of an address by one. Returns
          * positive it that succeeds, or 0 if this overflows. */
 
+        return in_addr_prefix_nth(family, u, prefixlen, 1);
+}
+
+/*
+ * Calculates the nth prefix of size prefixlen starting from the address demoted by u.
+ *
+ * On success 1 will be returned and the calculated prefix written to u. In
+ * case the calculation can not be performed (invalid prefix length, overflows
+ * would occur) 0  is returned.
+ * If the address family given isn't supported -EAFNOSUPPORT will be returned.
+ *
+ *
+ * Examples:
+ *   - in_addr_prefix_nth(AF_INET, 192.168.0.0, 24, 2), returns 1, writes 192.168.2.0 to u
+ *   - in_addr_prefix_nth(AF_INET, 255.255.255.0, 24, 1), returns 0, no data written
+ *   - in_addr_prefix_nth(AF_INET6, 2001:db8, 64, 0xff00) returns 1, writes 2001:0db8:0000:ff00:: to u
+ */
+int in_addr_prefix_nth(int family, union in_addr_union *u, unsigned prefixlen, uint64_t nth) {
+        assert(u);
+
         if (prefixlen <= 0)
                 return 0;
 
-        if (family == AF_INET) {
-                uint32_t c, n;
+        if (nth == 0)
+                return 1;
 
+        if (family == AF_INET) {
+                uint32_t c, n, t;
                 if (prefixlen > 32)
                         prefixlen = 32;
 
                 c = be32toh(u->in.s_addr);
-                n = c + (1UL << (32 - prefixlen));
-                if (n < c)
+
+                t = nth << (32 - prefixlen);
+
+
+                /* Check for wrap */
+                if (c >= t && UINT32_MAX - c < t)
+                        return 0;
+
+                if (t >= c && UINT32_MAX - t < c)
                         return 0;
```

Comment:
Are the two checks needed? IIUC, `if (c > UINT32_MAX - t)` should cover both cases.

Reply:
Yeah, you are right that is sufficient in this case. I'll change that.